polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -2343,64 +2342,49 @@ MLIRASTConsumer::GetOrCreateGlobal(const ValueDecl *FD, std::string prefix,
+  if (const Expr *Init = Var->getInit()) {
+    MLIRScanner MS(*this, module, LTInfo);
+    mlir::Block B;
+    // In case of device function, we will put the block in the forefront of
+    // the GPU module, else the block will go at the forefront of the main
+    // module.
+    if (FuncContext == FunctionContext::SYCLDevice) {
+      B.moveBefore(getDeviceModule(*module).getBody());
+      MS.getBuilder().setInsertionPointToStart(&B);
+    } else
+      MS.setEntryAndAllocBlock(&B);
 
-  if (tryInit)
-    if (auto init = VD->getInit()) {
-      MLIRScanner ms(*this, module, LTInfo);
-      mlir::Block *B = new Block();
-      // In case of device function, we will put the block in the forefront of
-      // the GPU module, else the block will go at the forefront of the main
-      // module.
-      if (funcContext == FunctionContext::SYCLDevice) {
-        B->moveBefore(getDeviceModule(*module).getBody());
-        ms.getBuilder().setInsertionPointToStart(B);
-      } else
-        ms.setEntryAndAllocBlock(B);
-      OpBuilder builder(module->getContext());
-      builder.setInsertionPointToEnd(B);
-      auto op = builder.create<memref::AllocaOp>(module->getLoc(), mr);
-
-      bool initialized = false;
-      if (isa<InitListExpr>(init)) {
-        if (auto A = ms.InitializeValueByInitListExpr(
-                op, const_cast<clang::Expr *>(init))) {
-          initialized = true;
-          initial_value = A;
-        }
-      } else {
-        auto VC = ms.Visit(const_cast<clang::Expr *>(init));
-        if (!VC.isReference) {
-          if (auto cop = VC.val.getDefiningOp<arith::ConstantOp>()) {
-            initial_value = cop.getValue();
-            initial_value = SplatElementsAttr::get(
-                RankedTensorType::get(mr.getShape(), mr.getElementType()),
-                initial_value);
-            initialized = true;
-          }
-        }
-      }
+    OpBuilder Builder(module->getContext());
+    Builder.setInsertionPointToEnd(&B);
+    auto Op = Builder.create<memref::AllocaOp>(module->getLoc(), VarTy);
 
-      if (!initialized) {
-        FD->dump();
-        init->dump();
-        llvm::errs() << " warning not initializing global: " << name << "\n";
-      } else {
-        globalOp.setInitialValueAttr(initial_value);
-      }
-      delete B;
+    if (isa<InitListExpr>(Init)) {
+      mlir::Attribute InitValAttr = MS.InitializeValueByInitListExpr(Op, Init);
+      globalOp.setInitialValueAttr(InitValAttr);
+    } else {
+      bool Initialized = false;
+      ValueCategory VC = MS.Visit(const_cast<clang::Expr *>(Init));
+      if (!VC.isReference)
+        if (auto Op = VC.val.getDefiningOp<arith::ConstantOp>()) {
+          auto InitialVal = SplatElementsAttr::get(
+              RankedTensorType::get(VarTy.getShape(), VarTy.getElementType()),
+              Op.getValue());
+          globalOp.setInitialValueAttr(InitialVal);
+          Initialized = true;
+        }
+      assert(Initialized);
```

Comment:
Why it must be initialized?

Reply:
can you please remove `Initialized` and change the two ifs to assert?

Reply:
done